File: HealthCareAtHome/shared/geo.js
/* geo.js — distance + eligibility helpers.
 *
 * Uses the haversine formula for straight-line (crow-flies) distance, which is
 * enough to decide which caregivers to alert. Precise road distance/ETA comes
 * from the Google Routes API in maps.js when a booking is actually accepted.
 */

const EARTH_RADIUS_KM = 6371;

function toRad(deg) {
  return (deg * Math.PI) / 180;
}

/** Great-circle distance in km between two {lat,lng} points. */
export function distanceKm(a, b) {
  if (!a || !b || a.lat == null || b.lat == null) return Infinity;
  const dLat = toRad(b.lat - a.lat);
  const dLng = toRad(b.lng - a.lng);
  const lat1 = toRad(a.lat);
  const lat2 = toRad(b.lat);
  const h =
    Math.sin(dLat / 2) ** 2 +
    Math.sin(dLng / 2) ** 2 * Math.cos(lat1) * Math.cos(lat2);
  return 2 * EARTH_RADIUS_KM * Math.asin(Math.sqrt(h));
}

/** Straight-line distance in metres between two {lat,lng} points (Infinity if unknown). */
export function distanceMeters(a, b) {
  const km = distanceKm(a, b);
  return isFinite(km) ? km * 1000 : Infinity;
}

/**
 * Location-verification policy check, shared by the client (start gate) and
 * caregiver (complete gate).
 *
 * @param {{lat,lng}} caregiverLoc  the caregiver's current/last-known location
 * @param {{lat,lng}} serviceLoc    the booking's service location
 * @param {{locationVerification:boolean, verifyRadiusMeters:number}} settings
 * @returns {{ ok:boolean, distanceMeters:number|null, reason:string }}
 *   - When verification is OFF: always ok (no check).
 *   - When ON: ok only if both locations have coords AND within the radius.
 *     Missing coordinates => not ok (fail-closed).
 */
export function checkProximity(caregiverLoc, serviceLoc, settings) {
  if (!settings || !settings.locationVerification) {
    return { ok: true, distanceMeters: null, reason: 'verification off' };
  }
  const radius = settings.verifyRadiusMeters || 50;
  const cHas = caregiverLoc && caregiverLoc.lat != null && caregiverLoc.lng != null;
  const sHas = serviceLoc && serviceLoc.lat != null && serviceLoc.lng != null;
  if (!cHas || !sHas) {
    return { ok: false, distanceMeters: null, reason: 'location unavailable' };
  }
  const d = distanceMeters(caregiverLoc, serviceLoc);
  return {
    ok: d <= radius,
    distanceMeters: Math.round(d),
    reason: d <= radius ? 'in range' : 'out of range'
  };
}

/** Rough ETA in minutes assuming an average urban speed (km/h). */
export function estimateEtaMinutes(distKm, avgSpeedKmh = 25) {
  if (!isFinite(distKm)) return null;
  return Math.max(1, Math.round((distKm / avgSpeedKmh) * 60));
}

/** True when a point has usable coordinates. */
function hasCoords(p) {
  return !!p && p.lat != null && p.lng != null;
}

/**
 * Filter caregivers eligible for a booking:
 *   - available
 *   - speciality matches the requested one
 *   - within `radiusKm` of the booking location — but ONLY when both the
 *     caregiver and the booking have real coordinates. If either location is
 *     missing coordinates (e.g. address not geocoded because no Maps key, or
 *     caregiver hasn't shared GPS), distance is unknown and we do NOT exclude
 *     on distance — the caregiver stays eligible so the request is still seen.
 * Returns caregivers annotated with `distanceKm` (Infinity when unknown),
 * nearest first (unknown-distance ones sorted last).
 */
/**
 * The caregiver location point(s) to compare against, per the admin match mode:
 *   'gps'        -> live shared location (cg.location)
 *   'registered' -> profile operating location (cg.operatingLocation)
 *   'both'       -> both points (eligible if EITHER is in range)
 */
export function caregiverMatchPoints(cg, mode = 'gps') {
  const pts = [];
  if (mode === 'registered') {
    if (hasCoords(cg.operatingLocation)) pts.push(cg.operatingLocation);
  } else if (mode === 'both') {
    if (hasCoords(cg.location)) pts.push(cg.location);
    if (hasCoords(cg.operatingLocation)) pts.push(cg.operatingLocation);
  } else { // 'gps'
    if (hasCoords(cg.location)) pts.push(cg.location);
  }
  return pts;
}

/** Smallest distance (km) from a caregiver to the booking, per match mode. */
export function caregiverDistanceKm(cg, booking, mode = 'gps') {
  const pts = caregiverMatchPoints(cg, mode);
  if (pts.length === 0) return Infinity;
  return Math.min(...pts.map((p) => distanceKm(p, booking.location)));
}

export function eligibleCaregivers(booking, caregivers, radiusKm, mode = 'gps') {
  const limit = radiusKm || booking.radiusKm || Infinity;
  const canMeasure = hasCoords(booking.location);
  return (caregivers || [])
    .filter((cg) => cg.availability === 'available')
    .filter((cg) => Array.isArray(cg.specialities) && cg.specialities.includes(booking.speciality))
    .map((cg) => ({ ...cg, distanceKm: caregiverDistanceKm(cg, booking, mode) }))
    .filter((cg) => {
      // apply the radius test only when we know the booking + at least one
      // caregiver point (per mode); otherwise keep eligible (distance unknown).
      const hasCgPoint = caregiverMatchPoints(cg, mode).length > 0;
      if (!canMeasure || !hasCgPoint) return true;
      return cg.distanceKm <= limit;
    })
    .sort((a, b) => a.distanceKm - b.distanceKm);
}

File: HealthCareFlutter/mobile/packages/hc_core/lib/src/util/geo.dart
import 'dart:math';

import '../models/enums.dart';
import '../models/models.dart';

/// Distance + eligibility helpers (ports the PWA's geo.js, including the fix
/// that the radius filter only applies when both points have coordinates).
class Geo {
  static const _earthRadiusKm = 6371.0;

  static double _toRad(double deg) => deg * pi / 180.0;

  /// Great-circle distance in km between two locations. Returns
  /// [double.infinity] when either lacks coordinates.
  static double distanceKm(HcLocation? a, HcLocation? b) {
    if (a == null || b == null || !a.hasCoords || !b.hasCoords) {
      return double.infinity;
    }
    final dLat = _toRad(b.lat! - a.lat!);
    final dLng = _toRad(b.lng! - a.lng!);
    final lat1 = _toRad(a.lat!);
    final lat2 = _toRad(b.lat!);
    final h = pow(sin(dLat / 2), 2) +
        pow(sin(dLng / 2), 2) * cos(lat1) * cos(lat2);
    return 2 * _earthRadiusKm * asin(sqrt(h));
  }

  /// Straight-line distance in metres (double.infinity when unknown).
  static double distanceMeters(HcLocation? a, HcLocation? b) {
    final km = distanceKm(a, b);
    return km.isFinite ? km * 1000 : double.infinity;
  }

  /// Rough ETA in minutes assuming an average urban speed (km/h).
  static int? estimateEtaMinutes(double distKm, [double avgSpeedKmh = 25]) {
    if (distKm.isInfinite || distKm.isNaN) return null;
    return max(1, (distKm / avgSpeedKmh * 60).round());
  }

  /// Location-verification policy check, shared by the client (start gate) and
  /// caregiver (complete gate). When verification is OFF -> always ok. When ON,
  /// ok only if both locations have coords AND within radius; missing coords is
  /// NOT ok (fail-closed).
  static ProximityVerdict checkProximity(
    HcLocation? caregiverLoc,
    HcLocation? serviceLoc,
    AppSettings settings,
  ) {
    if (!settings.locationVerification) {
      return const ProximityVerdict(ok: true, distanceMeters: null, reason: 'verification off');
    }
    final cHas = caregiverLoc != null && caregiverLoc.hasCoords;
    final sHas = serviceLoc != null && serviceLoc.hasCoords;
    if (!cHas || !sHas) {
      return const ProximityVerdict(ok: false, distanceMeters: null, reason: 'location unavailable');
    }
    final d = distanceMeters(caregiverLoc, serviceLoc);
    final within = d <= settings.verifyRadiusMeters;
    return ProximityVerdict(
      ok: within,
      distanceMeters: d.round(),
      reason: within ? 'in range' : 'out of range',
    );
  }

  /// The caregiver location point(s) to compare against, per the admin match
  /// mode: 'gps' -> live location; 'registered' -> operating location;
  /// 'both' -> both points (eligible if EITHER is in range).
  static List<HcLocation> caregiverMatchPoints(
    Caregiver cg, [
    String mode = MatchLocationMode.gps,
  ]) {
    final pts = <HcLocation>[];
    if (mode == MatchLocationMode.registered) {
      if (cg.operatingLocation != null && cg.operatingLocation!.hasCoords) {
        pts.add(cg.operatingLocation!);
      }
    } else if (mode == MatchLocationMode.both) {
      if (cg.location != null && cg.location!.hasCoords) pts.add(cg.location!);
      if (cg.operatingLocation != null && cg.operatingLocation!.hasCoords) {
        pts.add(cg.operatingLocation!);
      }
    } else {
      if (cg.location != null && cg.location!.hasCoords) pts.add(cg.location!);
    }
    return pts;
  }

  /// Smallest distance (km) from a caregiver to the booking, per match mode.
  static double caregiverDistanceKm(
    Caregiver cg,
    HcLocation bookingLocation, [
    String mode = MatchLocationMode.gps,
  ]) {
    final pts = caregiverMatchPoints(cg, mode);
    if (pts.isEmpty) return double.infinity;
    return pts
        .map((p) => distanceKm(p, bookingLocation))
        .reduce((a, b) => a < b ? a : b);
  }

  /// Filter caregivers eligible for a booking: available + speciality match +
  /// (when both have coords) within radius, per the match mode. Nearest-first.
  static List<Caregiver> eligibleCaregivers(
    Booking booking,
    List<Caregiver> caregivers,
    double? radiusKm, [
    String mode = MatchLocationMode.gps,
  ]) {
    final limit = radiusKm ?? booking.radiusKm ?? double.infinity;
    final canMeasure = booking.location.hasCoords;
    final matched = caregivers
        .where((cg) => cg.availability == Availability.available)
        .where((cg) => cg.specialities.contains(booking.speciality))
        .where((cg) {
      final hasPoint = caregiverMatchPoints(cg, mode).isNotEmpty;
      if (!canMeasure || !hasPoint) {
        return true; // distance unknown -> stay eligible
      }
      return caregiverDistanceKm(cg, booking.location, mode) <= limit;
    }).toList();

    matched.sort((a, b) => caregiverDistanceKm(a, booking.location, mode)
        .compareTo(caregiverDistanceKm(b, booking.location, mode)));
    return matched;
  }
}

/// Result of a location-verification proximity check.
class ProximityVerdict {
  final bool ok;
  final int? distanceMeters;
  final String reason;
  const ProximityVerdict({required this.ok, required this.distanceMeters, required this.reason});
}
